core/src/main/scala/com/github/johnynek/paiges/DocTree.scala

```diff
@@ -0,0 +1,330 @@
+    def split(x: Int): Option[(Option[Bounds], Bounds)] =
+      if (contains(x)) {
+        val left = if (x > min) Some(Bounds(min, x)) else None
+        Some((left, Bounds(x, max)))
+      }
+      else None
+  }
+
+  private[paiges] def toDocTree(d: Doc): DocTree = {
+
+    /**
+     * Return the minimum width needed to go down
+     * the left branch.
+     *
+     * Note we carry the current minimum (minV) in
+     * order to stop early.
+     */
+    @tailrec
+    def fits(pos: Int, d: DocTree, minV: Int): Int =
+      d.unfix match {
+        case Stream.Empty => pos min minV// we always can fit by going left
+        case Emit(Break(_)) #:: _ => pos min minV
+        case Emit(Str(s)) #:: tail =>
+          val nextPos = pos + s.length
+          if (nextPos >= minV) minV
+          else fits(nextPos, docTree(tail), minV)
+        case Split(a, b) #:: _ =>
+          val amin = cheatFits(pos, a, minV)
+          fits(pos, b(), amin)
+      }
+
+    def cheatFits(pos: Int, d: DocTree, minV: Int): Int =
+      fits(pos, d, minV)
+
+    @tailrec
+    def loop(pos: Int, lst: List[(Int, Doc)], bounds: Bounds): DocTree = lst match {
+      case Nil => docTree(Stream.empty)
+      case (i, Empty) :: z => loop(pos, z, bounds)
+      case (i, Concat(a, b)) :: z => loop(pos, (i, a) :: (i, b) :: z, bounds)
+      case (i, Nest(j, d)) :: z => loop(pos, ((i + j), d) :: z, bounds)
+      case (i, Text(s)) :: z => docTree(Emit(Str(s)) #:: cheat(pos + s.length, z, bounds).unfix)
+      case (i, Line) :: z => docTree(Emit(Break(i)) #:: cheat(i, z, bounds).unfix)
+      case (i, u@Union(a, _)) :: z =>
+        /**
+         * if we can go left, we do, otherwise we go right. So, in the current
+         * bounds, there is a threshold for the current node:
+         * if (w < wmin) go right
+         * else go left
+         */
+        val as = cheat(pos, (i, a) :: z, bounds)
+        val minLeftWidth = fits(pos, as, Int.MaxValue)
+        bounds.split(minLeftWidth) match {
+          case None =>
+            // cannot go left
+            loop(pos, (i, u.bDoc) :: z, bounds)
+          case Some((None, _)) =>
+            // always go left, because bounds.min == minLeftWidth
+            as
+          case Some((Some(rb), lb)) => // note when the width is smaller we go right
+            val left = cheat(pos, (i, a) :: z, lb)
+            def right = cheat(pos, (i, u.bDoc) :: z, rb)
+            docTree(Stream(Split(left, () => right)))
+        }
+    }
+
+    def cheat(pos: Int, lst: List[(Int, Doc)], bounds: Bounds): DocTree =
+      loop(pos, lst, bounds)
+
+    loop(0, (0, d) :: Nil, Bounds(0, Int.MaxValue))
+  }
+
+  private[paiges] def deunioned(d: DocTree): Stream[Doc] = {
+
+    def cat(a: Doc, b: Doc): Doc =
+      a match {
+        case Empty => b
+        case other => b match {
+          case Empty => other
+          case oo => Concat(other, oo)
+        }
+      }
+
+    @tailrec
+    def loop(tree: DocTree, prefix: Doc): Stream[Doc] =
```

Comment:
we could make a more efficient doc I think by doing: `prefix: List[Doc]` and the folding back from the right (which will be easy since it will be reversed, so foldLeft will do it) building back up the `Concat` which is totally right associated.